### PR TITLE
Implement Android single selection in CollectionView

### PR DESCRIPTION
**Pull request by hartez**
### Description of Change ###

Implement single item selection for the CollectionView on Android.

### Issues Resolved ### 

None

### API Changes ###
Added:

- SelectionMode SelectableItemsView.SelectionMode
- object SelectableItemsView.SelectedItem
- ICommand SelectableItemsView.SelectionChangedCommand
- object SelectableItemsView.SelectionChangedCommandParameter
- event EventHandler<SelectionChangedEventArgs> SelectionChanged
- enum SelectionMode 
- class SelectionChangedEventArgs

### Platforms Affected ### 

- Core/XAML (all platforms)
- Android

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### Testing Procedure ###

In Control Gallery, navigate to CollectionView Gallery -> Selection Galleries

### PR Checklist ###

- [ ] Has automated tests 
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

File: Xamarin.Forms.Core/Items/ItemsView.cs [hunk 1]
```diff
@@ -97,4 +98,29 @@ protected virtual void OnScrollToRequested(ScrollToRequestEventArgs e)
...
+	}
+
+	public class SelectionChangedEventArgs : EventArgs
+	{
+		public IReadOnlyList<object> PreviousSelection { get; }
+		public IReadOnlyList<object> CurrentSelection { get; }
+
+		public SelectionChangedEventArgs(object previousSelection, object currentSelection)
+		{
+			PreviousSelection = previousSelection != null ? new List<object> { previousSelection } : new List<object>();
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Items/ItemsView.cs]: - the list will never grows, construct as `new List<object> (1) { previousSelection}`
- if the empty case happens often, you might want to have an static `EmptyList` that you can reuse

File: Xamarin.Forms.Core/Items/ItemsView.cs [hunk 2]
```diff
@@ -97,4 +98,29 @@ protected virtual void OnScrollToRequested(ScrollToRequestEventArgs e)
...
+		Single,
+		Multiple
+	}
+
+	public class SelectionChangedEventArgs : EventArgs
+	{
+		public IReadOnlyList<object> PreviousSelection { get; }
+		public IReadOnlyList<object> CurrentSelection { get; }
+
+		public SelectionChangedEventArgs(object previousSelection, object currentSelection)
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Items/ItemsView.cs]: as this one is a convenience helper, I'd make it internal

File: Xamarin.Forms.Core/Items/SelectableItemsView.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+		public static readonly BindableProperty SelectionChangedCommandParameterProperty =
+			BindableProperty.Create(nameof(SelectionChangedCommandParameter), typeof(object),
+				typeof(SelectableItemsView));
+
+		public object SelectedItem
+		{
+			get => GetValue(SelectedItemProperty);
+			set => SetValue(SelectedItemProperty, value);
+		}
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Items/SelectableItemsView.cs]: for a braces lover, that's a surprisingly low number of lines and braces for a single property 🤪

File: Xamarin.Forms.Core/Items/SelectableItemsView.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+			{
+				var commandParameter = SelectionChangedCommandParameter;
+
+				if (command.CanExecute(commandParameter))
+				{
+					command.Execute(commandParameter);
+				}
+			}
+			
+			SelectionChanged?.Invoke(this, args);
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Items/SelectableItemsView.cs]: ideally, we should avoid triggering events from virtual methods

File: Xamarin.Forms.Core/Items/SelectableItemsView.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+		}
+
+		static void SelectedItemPropertyChanged(BindableObject bindable, object oldValue, object newValue)
+		{
+			var args = new SelectionChangedEventArgs(oldValue, newValue);
+			((SelectableItemsView)bindable).OnSelectionChanged(args);
+		}
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Items/SelectableItemsView.cs]: and we should fire the event from here

Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/Items/SelectableItemsView.cs]: actually, we should fire the event, and trigger the command from here
leaving the virtual method completely empty...

File: Xamarin.Forms.Core/Items/ItemsView.cs (hunk 1 above)
Comment (hartez) [reply Xamarin.Forms.Core/Items/ItemsView.cs]: Done and done.

File: Xamarin.Forms.Core/Items/ItemsView.cs (hunk 2 above)
Comment (hartez) [reply Xamarin.Forms.Core/Items/ItemsView.cs]: > as this one is a convenience helper, I'd make it internal

Done.

>  move SelectionMode enum and SelectionChangedEventArgs class

Moved them to their own files.

File: Xamarin.Forms.Core/Items/SelectableItemsView.cs (hunk 4 above)
Comment (hartez) [reply Xamarin.Forms.Core/Items/SelectableItemsView.cs]: Fixed.

File: Xamarin.Forms.Core/Items/SelectableItemsView.cs (hunk 5 above)
Comment (hartez) [reply Xamarin.Forms.Core/Items/SelectableItemsView.cs]: Fixed - now fires the command and the event from the handler, and the virtual method is empty.

**Comment by StephaneDelcroix** [comment]
👍 for the latest changes

File: Xamarin.Forms.Core/Items/SelectableItemsView.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.Windows.Input;
+
+namespace Xamarin.Forms
+{
+	public class SelectableItemsView : ItemsView
+	{
+		public static readonly BindableProperty SelectionModeProperty =
+			BindableProperty.Create(nameof(SelectionMode), typeof(SelectionMode), typeof(SelectableItemsView),
+				SelectionMode.None);
```
[the remark this answers is not in the document]
Comment (hartez) [reply Xamarin.Forms.Core/Items/SelectableItemsView.cs]: @paymicro @StephaneDelcroix I'm not sure we should be so prescriptive here.

How should the control behave if the SelectionMode is 'None' and the user sets the SelectedItem?

File: Xamarin.Forms.Core/Items/SelectableItemsView.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+		public static readonly BindableProperty SelectedItemProperty =
+			BindableProperty.Create(nameof(SelectedItem), typeof(object), typeof(SelectableItemsView),
```
[the remark this answers is not in the document]
Comment (hartez) [reply Xamarin.Forms.Core/Items/SelectableItemsView.cs]: Done.

File: Xamarin.Forms.Platform.Android/CollectionView/SelectableItemsViewRenderer.cs [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+			SwapAdapter(ItemsViewAdapter, true);
+		}
+
+		void ClearSelection()
+		{
+			for (int i = 0, size = ChildCount; i < size; i++)
+			{
+				var holder = GetChildViewHolder(GetChildAt(i));
+				if (holder != null)
+				{
```
[the remark this answers is not in the document]
Comment (hartez) [reply Xamarin.Forms.Platform.Android/CollectionView/SelectableItemsViewRenderer.cs]: Removed.

File: Xamarin.Forms.Controls/Xamarin.Forms.Controls.csproj [hunk 9]
```diff
@@ -48,7 +48,92 @@
     <PackageReference Include="Newtonsoft.Json" Version="10.0.3" />
     <PackageReference Include="Xam.Plugin.DeviceInfo" Version="3.0.2" />
   </ItemGroup>
-
+  <ItemGroup>
```
[the remark this answers is not in the document]
Comment (hartez) [reply Xamarin.Forms.Controls/Xamarin.Forms.Controls.csproj]: No clue how it got in there. Maybe a rebase issue? Or just VS behaving badly. Anyway, it's gone now.

File: Xamarin.Forms.Core/Items/SelectableItemsView.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (paymicro) [reply Xamarin.Forms.Core/Items/SelectableItemsView.cs]: > How should the control behave if the SelectionMode is 'None' and the user sets the SelectedItem?
I think that in this case nothing needs to be done.

[the remark this answers is not in the document]
Comment (hartez) [reply Xamarin.Forms.Core/Items/SelectableItemsView.cs]: > I think that in this case nothing needs to be done.

I feel like that would be confusing for a user doing something like this:
`<CollectionView SelectedItem="{Binding VMSelectedItem}" SelectionMode="Single"/>`

If the property setter for `SelectedItem` is called before the property setter for `SelectionMode`, then the `SelectedItem` property will not be set.

[the remark this answers is not in the document]
Comment (hartez) [reply Xamarin.Forms.Core/Items/SelectableItemsView.cs]: And we'd have a similar problem if the user were trying something like 

`<CollectionView SelectedItems="{Binding VMSelectedItems}" SelectionMode="{Binding VMSelectionMode}" />`


The order of resolution of those bindings changes the outcome. 

If the user wants the behavior you describe [upthread](https://github.com/xamarin/Xamarin.Forms/pull/4651#discussion_r239815360), they can implement that easily themselves in a custom class, but we could make it easier by adding a virtual `OnSelectionModeChanged` method.

[the remark this answers is not in the document]
Comment (paymicro) [reply Xamarin.Forms.Core/Items/SelectableItemsView.cs]: Ok you convinced me. if it creates more problems than it solves.

**Pull request merged**